File: scripts/sv_logger.js
//Requires
const utils = require('./utils.js')

//Helpers
const log = (x)=>{console.log(`^5[txAdminClientJS]^0 ${x}`)}
const dir = (x)=>{console.dir(x)}
const isUndefined = (x) => { return (typeof x === 'undefined') };
const logError = (err) => {
    if(typeof err === 'string'){
        console.log(`^5[txAdminClient:Error]^1 ${err}^0`);
    }else{
        console.log(`^5[txAdminClient:Error]^1 ${err.message}^0`);
        try {
            err.stack.forEach(trace => {
                console.log(`\t=> ${trace.file}:${trace.line} > ${trace.name}`)
            });
        } catch (error) {
            console.log('Error stack unavailable.')
        }
        console.log()
    }
}
const getIdentifiers = (src) => {
    let identifiers = [];
    try {
        for (let i = 0; i < GetNumPlayerIdentifiers(src); i++) {
            identifiers.push(GetPlayerIdentifier(src, i))
        }
    } catch (error) {
        logError(error)
    }
    return identifiers;
}
const getPlayerData = (src) => {
    if(src === null || src === false) return false;
    if(src === -1) return {name: 'console', identifiers:[]}
    try {
        return {
            name: GetPlayerName(src),
            identifiers: getIdentifiers(src)
        };
    } catch (error) {
        logError(error)
        return false;
    }
}
const debugPrint = (data) => {
    let sep = '='.repeat(46);
    let json = JSON.stringify(data, null, 2);
    log(`${sep}\n${json}\n${sep}`);
}


/**
 * Logger class
 */
class Logger {
    constructor(){
        log('Logger started');
        this.log = [{
            timestamp: (Date.now() / 1000).toFixed(),
            action: "txAdminClient:Started",
            source: false,
            data: false
        }];

        //Attempt to set env vars
        this.txAdminPort = GetConvar("txAdmin-apiPort", "invalid");
        this.txAdminToken = GetConvar("txAdmin-apiToken", "invalid");
        if(this.txAdminPort === 'invalid' || this.txAdminToken === 'invalid'){
            logError('API Port and Token ConVars not found. Do not start this resource if not using txAdmin.') 
        }

        //Attempt to flush log to txAdmin, starting after 10 seconds
        setTimeout(() => {
            setInterval(() => {
                this.flushLog();
            }, 2500);
        }, 10000);
    }

    //Register log event
    r(src, action, data){
        let toLog = {
            timestamp: (Date.now() / 1000).toFixed(),
            action,
            source: getPlayerData(src),
            data: (data)? data : false
        }
        this.log.push(toLog)
        // debugPrint(toLog);
    }

    //Flush Log
    flushLog(){
        if(!this.log.length) return;

        const postData = JSON.stringify({
            txAdminToken: this.txAdminToken,
            log: this.log
        })
        utils.postJson(`http://localhost:${this.txAdminPort}/intercom/logger`, postData)
            .then((data) => {
                if(data.statusCode === 413){
                    log(`Logger upload failed with code 413 and body ${data.body}`);
                    //TODO: introduce a buffer to re-upload the log in parts.
                    this.log = [{
                        timestamp: (Date.now() / 1000).toFixed(),
                        action: "DebugMessage",
                        source: false,
                        data: `wiped log with size ${postData.length} due to upload limit`
                    }];
                }else if(data.statusCode === 200){
                    this.log = [];
                }else{
                    log(`Logger upload failed with code ${data.statusCode} and body ${data.body}`);
                }
            })
            .catch((error) => {
                log(`Logger upload failed with error: ${error.message}`);
            });
    }
}
logger = new Logger();


//Event handlers
on('playerConnecting', (name, skr, d) => {
    try {
        logger.r(global.source, 'playerConnecting');
    } catch (error) {
        logError(error)
    }
})

on('playerDropped', (reason) => {
    try {
        logger.r(global.source, 'playerDropped');
    } catch (error) {
        logError(error)
    }
})

on('explosionEvent', (source, ev) => {
    //Helper function
    const isInvalid = (prop, invValue) => {
        return (typeof prop == 'undefined' || prop === invValue);
    }
    //Filtering out bad event calls
    if(
        isInvalid(ev.damageScale, 0) ||
        isInvalid(ev.cameraShake, 0) ||
        isInvalid(ev.isInvisible, true) ||
        isInvalid(ev.isAudible, false)
    ){
        return;
    }
    //Decoding explosion type
    let types = ['DONTCARE', 'GRENADE', 'GRENADELAUNCHER', 'STICKYBOMB', 'MOLOTOV', 'ROCKET', 'TANKSHELL', 'HI_OCTANE', 'CAR', 'PLANE', 'PETROL_PUMP', 'BIKE', 'DIR_STEAM', 'DIR_FLAME', 'DIR_WATER_HYDRANT', 'DIR_GAS_CANISTER', 'BOAT', 'SHIP_DESTROY', 'TRUCK', 'BULLET', 'SMOKEGRENADELAUNCHER', 'SMOKEGRENADE', 'BZGAS', 'FLARE', 'GAS_CANISTER', 'EXTINGUISHER', 'PROGRAMMABLEAR', 'TRAIN', 'BARREL', 'PROPANE', 'BLIMP', 'DIR_FLAME_EXPLODE', 'TANKER', 'PLANE_ROCKET', 'VEHICLE_BULLET', 'GAS_TANK', 'BIRD_CRAP', 'RAILGUN', 'BLIMP2', 'FIREWORK', 'SNOWBALL', 'PROXMINE', 'VALKYRIE_CANNON', 'AIR_DEFENCE', 'PIPEBOMB', 'VEHICLEMINE', 'EXPLOSIVEAMMO', 'APCSHELL', 'BOMB_CLUSTER', 'BOMB_GAS', 'BOMB_INCENDIARY', 'BOMB_STANDARD', 'TORPEDO', 'TORPEDO_UNDERWATER', 'BOMBUSHKA_CANNON', 'BOMB_CLUSTER_SECONDARY', 'HUNTER_BARRAGE', 'HUNTER_CANNON', 'ROGUE_CANNON', 'MINE_UNDERWATER', 'ORBITAL_CANNON', 'BOMB_STANDARD_WIDE', 'EXPLOSIVEAMMO_SHOTGUN', 'OPPRESSOR2_CANNON', 'MORTAR_KINETIC', 'VEHICLEMINE_KINETIC', 'VEHICLEMINE_EMP', 'VEHICLEMINE_SPIKE', 'VEHICLEMINE_SLICK', 'VEHICLEMINE_TAR', 'SCRIPT_DRONE', 'RAYGUN', 'BURIEDMINE', 'SCRIPT_MISSIL'];
    if(ev.explosionType < -1 || ev.explosionType > 72){
        ev.explosionType = 'UNKNOWN';
    }else{
        ev.explosionType =  types[ev.explosionType+1]
    }
    //Adding logging data
    try {
        logger.r(source, 'explosionEvent', ev);
    } catch (error) {
        logError(error)
    }
})

onNet('txaLogger:DeathNotice', (killer, cause) => {
    let killerData = null;
    if(killer !== null && killer !== false){
        try {
            killerData = getPlayerData(GetPlayerFromIndex(killer));
        } catch (error) {}
    }
    try {
        let toLogData = {
            cause,
            killer: killerData
        }
        logger.r(global.source, 'DeathNotice', toLogData);
    } catch (error) {
        logError(error)
    }
});

onNet('txaLogger:CommandExecuted', (data) => {
    try {
        logger.r(global.source, 'CommandExecuted', data);
    } catch (error) {
        logError(error)
    }
});

onNet('txaLogger:DebugMessage', (data) => {
    try {
        logger.r(global.source, 'DebugMessage', data);
    } catch (error) {
        logError(error)
    }
});

const logChatMessage = (src, author, text)=>{
    try {
        let toLogData = {
            author,
            text
        }
        logger.r(src, 'ChatMessage', toLogData);
    } catch (error) {
        logError(error)
    }
}
onNet('chatMessage', logChatMessage);
onNet('txaLogger:internalChatMessage', logChatMessage);

File: scripts/sv_logger.lua
--Check Environment
if GetConvar('txAdminServerMode', 'false') ~= 'true' then
    return
end

-- micro optimization
local os_time = os.time
-- http://lua-users.org/wiki/SimpleRound
local function round(num)
    return tonumber(string.format("%.2f", num))
end

local function getPlayerData(src)
    if type(src) == 'string' then
        src = tonumber(src)
    end

    if not src then
        return false
    end

    if src <= 0 then return {name = 'console', identifiers = {}} end

    return {
        name = GetPlayerName(src),
        identifiers = GetPlayerIdentifiers(src)
    }
end

local loggerBuffer = {}
local PRINT_STRUCTURED_TRACE = `PRINT_STRUCTURED_TRACE` & 0xFFFFFFFF
--- function logger
--- Sends logs through fd3 to the server & displays the logs on the panel.
---@param src number the source of the player who did the action
---@param action string the action type
---@param data table|boolean will take a table, or a boolean if there is no data.
local function logger(src, action, data)
    loggerBuffer[#loggerBuffer+1] = {
        timestamp = round(os_time()),
        source = getPlayerData(src),
        action = action,
        data = data or false
    }
end

-- send all of the buffered logs every second
CreateThread(function()
    while true do
        Wait(1000)
        if #loggerBuffer > 0 then
            local payload = json.encode({
                type = 'txAdminLogData',
                logs = loggerBuffer
            })
            Citizen.InvokeNative(PRINT_STRUCTURED_TRACE, payload)
            loggerBuffer = {}
        end
    end
end)

logger(-1, 'txAdminClient:Started')

AddEventHandler('playerConnecting', function()
    logger(source, 'playerConnecting')
end)

-- RegisterNetEvent('playerJoining', function()
--     logger(source, 'playerJoining')
-- end)

AddEventHandler('playerDropped', function()
    logger(source, 'playerDropped')
end)

local function isInvalid(property, invalidType)
    return (property == nil or property == invalidType)
end

local explosionTypes = {'GRENADE', 'GRENADELAUNCHER', 'STICKYBOMB', 'MOLOTOV', 'ROCKET', 'TANKSHELL', 'HI_OCTANE', 'CAR', 'PLANE', 'PETROL_PUMP', 'BIKE', 'DIR_STEAM', 'DIR_FLAME', 'DIR_WATER_HYDRANT', 'DIR_GAS_CANISTER', 'BOAT', 'SHIP_DESTROY', 'TRUCK', 'BULLET', 'SMOKEGRENADELAUNCHER', 'SMOKEGRENADE', 'BZGAS', 'FLARE', 'GAS_CANISTER', 'EXTINGUISHER', 'PROGRAMMABLEAR', 'TRAIN', 'BARREL', 'PROPANE', 'BLIMP', 'DIR_FLAME_EXPLODE', 'TANKER', 'PLANE_ROCKET', 'VEHICLE_BULLET', 'GAS_TANK', 'BIRD_CRAP', 'RAILGUN', 'BLIMP2', 'FIREWORK', 'SNOWBALL', 'PROXMINE', 'VALKYRIE_CANNON', 'AIR_DEFENCE', 'PIPEBOMB', 'VEHICLEMINE', 'EXPLOSIVEAMMO', 'APCSHELL', 'BOMB_CLUSTER', 'BOMB_GAS', 'BOMB_INCENDIARY', 'BOMB_STANDARD', 'TORPEDO', 'TORPEDO_UNDERWATER', 'BOMBUSHKA_CANNON', 'BOMB_CLUSTER_SECONDARY', 'HUNTER_BARRAGE', 'HUNTER_CANNON', 'ROGUE_CANNON', 'MINE_UNDERWATER', 'ORBITAL_CANNON', 'BOMB_STANDARD_WIDE', 'EXPLOSIVEAMMO_SHOTGUN', 'OPPRESSOR2_CANNON', 'MORTAR_KINETIC', 'VEHICLEMINE_KINETIC', 'VEHICLEMINE_EMP', 'VEHICLEMINE_SPIKE', 'VEHICLEMINE_SLICK', 'VEHICLEMINE_TAR', 'SCRIPT_DRONE', 'RAYGUN', 'BURIEDMINE', 'SCRIPT_MISSIL'}

AddEventHandler('explosionEvent', function(source, ev)
    if (isInvalid(ev.damageScale, 0) or isInvalid(ev.cameraShake, 0) or isInvalid(ev.isInvisible, true) or
        isInvalid(ev.isAudible, false)) then
        return
    end

    if ev.explosionType < -1 or ev.explosionType > 77 then
        ev.explosionType = 'UNKNOWN'
    else
        ev.explosionType = explosionTypes[ev.explosionType + 1]
    end

    logger(source, 'explosionEvent', ev)
end)

-- An internal server handler, this is NOT exposed to the client
AddEventHandler('txaLogger:menuEvent', function(source, event, allowed, data)
    local message
    if event == 'healSelf' then
        message = "healing themself"

    elseif event == 'healAll' then
        message = "healing all players!"

    elseif event == 'teleportCoords' then
        if type(data) ~= 'table' then return end
        local x = data.x
        local y = data.y
        local z = data.z
        message = ("teleporting to coordinates (x=%.3f, y=%0.3f, z=%0.3f)"):format(x or 0.0, y or 0.0, z or 0.0)

    elseif event == 'teleportWaypoint' then
        message = "teleporting to a waypoint"

    elseif event == 'announcement' then
        if type(data) ~= 'string' then return end
        message = "making a server-wide announcement: " .. data

    elseif event == 'vehicleRepair' then
        message = "repairing their vehicle"

    elseif event == 'spawnVehicle' then
        if type(data) ~= 'string' then return end
        message = "spawning a vehicle (model: " .. data .. ")"

    elseif event == 'playerModeChanged' then
        if data == 'godmode' then
            message = "enabling invincibility"
        elseif data == 'noclip' then
            message = "enabling noclip"
        elseif data == 'none' then
            message = "becoming mortal (standard mode)"
        else
            message = "invalid player mode"
        end

    elseif event == 'teleportPlayer' then
        if type(data) ~= 'table' then return end
        local playerName = data.playerName
        if type(playerName) ~= 'string' then return end
        local x = data.x or 0.0
        local y = data.y or 0.0
        local z = data.z or 0.0
        message = ("teleporting to player %s (x=%.3f, y=%.3f, z=%.3f)"):format(playerName, x, y, z)

    elseif event == 'healPlayer' then
        if type(data) ~= 'string' then return end
        message = "healing player " .. data

    elseif event == 'summonPlayer' then
        if type(data) ~= 'string' then return end
        message = "summoning player " .. data

    elseif event == 'weedEffect' then
        if type(data) ~= 'string' then return end
        message = "triggering weed effect on " .. data

    elseif event == 'drunkEffect' then
        if type(data) ~= 'string' then return end
        message = "triggering drunk effect on " .. data

    elseif event == 'wildAttack' then
        if type(data) ~= 'string' then return end
        message = "triggering wild attack on " .. data

    elseif event == 'setOnFire' then
        if type(data) ~= 'string' then return end
        message = "setting ".. data .." on fire" 

    elseif event == 'clearArea' then
        if type(data) ~= 'number' then return end
        message = "clearing an area with ".. data .."m radius"

    else
        logger(source, 'DebugMessage', "unknown menu event "..event)
        return
    end
    
    local event_data = { message = message, allowed = allowed }
    logger(source, 'MenuEvent', event_data)
end)

RegisterNetEvent('txaLogger:DeathNotice', function(killer, cause)
    local killerData
    if killer then
        killerData = getPlayerData(killer)
    end
    local logData = {
        cause = cause,
        killer = killerData
    }
    logger(source, 'DeathNotice', logData)
end)

RegisterNetEvent('txaLogger:CommandExecuted', function(data)
    logger(source, 'CommandExecuted', data)
end)

RegisterNetEvent('txaLogger:DebugMessage', function(data)
    logger(source, 'DebugMessage', data)
end)

local function logChatMessage(src, author, text)
    local logData = {
        author = author,
        text = text
    }
    logger(src, 'ChatMessage', logData)
end
RegisterNetEvent('chatMessage', logChatMessage)
RegisterNetEvent('txaLogger:internalChatMessage', logChatMessage)
